### PR TITLE
s3d: network-level diagnostics toolkit (oracle ladder, bootstrap CIs, probes, gradient conflict)

### PR DESCRIPTION
**Pull request by Bovey0809**
Implements `docs/superpowers/specs/2026-07-21-s3d-network-diagnostics-design.md` — a repeatable analysis method that, given any s3d checkpoint + dataset, attributes AP3D loss to network components and ranks the next levers with quantified headroom.

## What's included

- **`ultralytics/models/yolo/s3d/diagnose.py`**: pred/GT matching, per-component signed error records, depth bias/variance fit (`dz = a·z + b`), MonoDLE/TIDE3D **oracle-swap ladder** (z / y / xy / location / dims / orientation / **score-ranking** oracles) with recomputed rotated IoU, **bootstrap AP CIs** (metric-noise floor for small val sets), distribution-shape probes (softmax entropy/top-2 for DepthDFL bins and cost-volume disparity slices), **counterfactual stereo-shift math** (is depth stereo-driven or a monocular shortcut?), and a **per-branch gradient probe** (GradNorm norms + PCGrad cosine conflict on the shared neck).
- **`ultralytics/data/scripts/diagnose_s3d.py`**: one-pass CLI (pattern: `benchmark_s3d.py`) that drives the standard validator with probe hooks attached and emits `report.md` + CSVs with a ranked lever list.
- **`Stereo3DDetLoss.loss()` construction relocated into `component_losses()`** — single loss-construction path; `loss()` is now a thin stack-and-scale wrapper. Exact-equality test guards the refactor.
- **Opt-in gradient callback**: model YAML `training.diag_grad_every: N` appends per-branch gradient norms + pairwise cosines to `grad_diag.csv` every N batches via `torch.autograd.grad` (never touches `.grad`). Zero behavior change when unset.

## Testing

- 19 new CPU unit tests in `tests/test_s3d_diagnose.py` (synthetic Box3D stats, toy-model gradient conflict, loss-refactor equality); full non-network s3d suite passes (43 tests).
- End-to-end GPU validation on the client cube_s3d dataset (RTX PRO 6000): val pass reproduces the known cv48 checkpoint metrics; all report sections (oracle ladder, bootstrap CIs, DFL/cost-volume probes, stereo-shift, gradient conflict matrix) produce data.

Deleted: nothing removed in-repo beyond the `loss()` body relocation (net-negative in loss.py); consolidates and supersedes the off-repo profiling scripts (diag8.py / profile_arm.py / profile_ema.py / farm_s3d_collect.py per-component sections) — reuses `Stereo3DDetMetrics`, `compute_3d_iou_batch`, `decode_and_refine_predictions` rather than adding parallel implementations. Rule-2 justification: error attribution/probing is net-new capability, not expressible as deletion.

Known follow-up (documented in spec): a decode-space depth oracle (overwriting `outputs["depth"]` pre-decode) to separate representation error from fusion error; the box-space `z` oracle here answers the headroom question.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🧭 Adds a comprehensive S3D network diagnostics toolkit for identifying stereo 3D detection errors, performance bottlenecks, and training issues.

### 📊 Key Changes
- Added `diagnose_s3d` analysis utilities for:
  - Per-object error records and depth-bias profiling.
  - Oracle component swaps to estimate AP headroom for depth, location, dimensions, orientation, and confidence ranking.
  - Bootstrap AP confidence intervals.
  - DepthDFL and stereo cost-volume distribution analysis.
  - Counterfactual stereo sensitivity testing.
  - Per-branch gradient magnitude and conflict diagnostics.
- Added a command-line diagnostic script that runs validation once and generates:
  - A Markdown `report.md`.
  - CSV files containing error and probe records.
  - Ranked recommendations for the model components with the greatest potential impact.
- Refactored S3D loss computation with `component_losses()` so individual loss terms remain connected to the autograd graph for gradient analysis.
- Added optional training-time gradient diagnostics through the `training.diag_grad_every` configuration.
- Added extensive unit tests covering matching, error summaries, oracle swaps, bootstrap metrics, stereo geometry, distribution statistics, gradients, and report generation.
- Added API reference pages and navigation entries for the new diagnostic modules.

### 🎯 Purpose & Impact
- 🛠️ Helps developers quickly determine whether accuracy issues come from depth estimation, stereo fusion, localization, dimensions, orientation, ranking, or optimization conflicts.
- 📈 Provides more reliable evaluation context by exposing metric uncertainty on small validation sets instead of relying on AP alone.
- 🔍 Makes stereo-specific failure modes easier to detect, including monocular shortcuts, depth scale bias, poor cost-volume quality, and conflicting loss branches.
- 🚀 Produces actionable reports and ranked improvement opportunities, potentially reducing S3D model debugging and iteration time.
- ✅ The core training and validation behavior remains consistent, while diagnostics are opt-in through the new script and configuration setting.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated review message, and an engineer will assist with any follow-up questions. Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The oracle, bootstrap, and loss-component refactor are well covered by focused tests. However, the training gradient probe is currently unreachable, and two probe calculations report incorrect values under normal model configurations.

💬 Posted 3 inline comments

File: ultralytics/models/yolo/s3d/diagnose.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+    """
+    from ultralytics.utils.torch_utils import unwrap_model
+
+    every = getattr(trainer, "_diag_grad_every", 0)
+    ib = getattr(trainer, "_diag_batch_i", 0)
+    trainer._diag_batch_i = ib + 1
+    if not every or ib % every:
+        return
+    model = unwrap_model(trainer.model)
+    if getattr(model, "criterion", None) is None or not hasattr(trainer, "batch"):
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/s3d/diagnose.py:405]: ⚠️ **HIGH**: `BaseTrainer._do_train()` keeps the preprocessed batch in a local `batch` variable and never assigns `trainer.batch`. Therefore this guard always returns during normal training, so setting `training.diag_grad_every` silently produces no gradient diagnostics or CSV output. Store the current batch on the trainer before this callback, or capture/pass it through the callback registration.

File: ultralytics/data/scripts/diagnose_s3d.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                    row.update(dfl_entropy=float(ds["entropy"][0]), dfl_top2=float(ds["top2_mass"][0]))
+            if cost_vol is not None:
+                stride_cv = in_h / cost_vol.shape[2]
+                cu, cv_ = int(u / stride_cv), int(v / stride_cv)
+                if 0 <= cv_ < cost_vol.shape[2] and 0 <= cu < cost_vol.shape[3]:
+                    ds = dist_stats(cost_vol[si, :, cv_, cu][None])
+                    gt_disp_bins = (calib["fx"] * scale) * calib["baseline"] / z / stride_cv
+                    row.update(
+                        cv_entropy=float(ds["entropy"][0]),
+                        cv_peak_err_bins=float(abs(float(ds["argmax"][0]) - gt_disp_bins)),
```
Comment (UltralyticsAssistant) [inline ultralytics/data/scripts/diagnose_s3d.py:183]: 💡 **MEDIUM**: This compares the cost-volume argmax *index* with a disparity measured in feature pixels. `StereoCostVolume` samples its `disparities` offsets from 0 to `max_disp`; with the current 24-bin/48-pixel configuration, index 6 represents an offset near 12 pixels, not 6. As a result `cv_peak_err_bins` and the report's median peak error are numerically incorrect. Convert the GT feature disparity to the nearest `StereoCostVolume.disparities` index, or compare the argmax's mapped offset to the GT disparity.

File: ultralytics/data/scripts/diagnose_s3d.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+                probe_rows.extend(_gt_probe_rows(v, captured))
+                if v.batch_i < args.shift_batches:
+                    shift_pairs.extend(_shift_probe(model.model, v, args.shift_px))
+                    if fx0[0] is None and v._current_batch.get("calib"):
+                        c0 = v._current_batch["calib"][0]
+                        ori0 = v._current_batch["ori_shape"][0]
+                        from ultralytics.models.yolo.s3d.preprocess import compute_letterbox_params
+
+                        s0, _, _ = compute_letterbox_params(int(ori0[0]), int(ori0[1]), v.args.imgsz)
+                        fx0[0] = float(c0["fx"]) / s0  # original-image fx
```
Comment (UltralyticsAssistant) [inline ultralytics/data/scripts/diagnose_s3d.py:332]: 💡 **MEDIUM**: `shift_px` is applied to the letterboxed tensor, so its disparity change is in letterbox pixels. `c0["fx"]` is also already in letterbox space after `StereoLetterBox`, but this divides it back to original-image space before passing it with the unscaled shift. For any non-unit resize scale, the reported stereo sensitivity is biased by that scale. Keep the letterbox-space focal length here (or scale `dpx` into original-image pixels).

**Suggested change:**
```suggestion
                        fx0[0] = float(c0["fx"])
```

**Comment by Bovey0809** [comment]
Closing — internal tooling, not for upstream.
